addons/test_mail/tests/test_mail_composer.py

```diff
@@ -1852,6 +1859,164 @@ def test_mail_composer_delete(self):
         self.assertEqual(len(self._new_msgs), 2, 'Should have created 1 mail.mail per record')
         self.assertFalse(self._new_msgs.exists(), 'Should have deleted mail.message records')
 
+    @users('employee')
+    @mute_logger('odoo.tests', 'odoo.addons.mail.models.mail_mail', 'odoo.models.unlink')
+    def test_mail_composer_duplicates(self):
+        """ Ensures emails sent to the same recipient multiple times
+            are only sent when they are not duplicates
+        """
+        self.template.write({
+            'auto_delete': False,
+            'body_html': '<p>Common Body</p>',
+            'subject': 'Common Subject',
+            'partner_to': '{{ object.customer_id.id if object.customer_id else "" }}',
+            'email_to': ''
+        })
+
+        # Guarantee points of variation for test_report_3
+        self.test_records[0].write({'count': 1, 'name': 'A'})
+        self.test_records[1].write({'count': 2, 'name': 'B'})
+
+        template_attachment, composer_attachment = self.env['ir.attachment'].create([{
+            'datas': base64.b64encode(b'ExtraData'),
+            'mimetype': 'text/plain',
+            'name': f'{record._name}_Common_Attachment.txt',
+            'res_model': record._name,
+            'res_id': record.id if record else False,
+        } for record in (self.template, self.env['mail.compose.message'])])
+
+        customer_ids = self.env['res.partner'].create([{
```

Comment:
I think `_create_records_for_batch` already creates customer for self.test_records, can't you update records instead of creating new partners ?